matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/membership/leaving/LeaveRoomTask.kt

```diff
@@ -31,12 +40,40 @@ internal interface LeaveRoomTask : Task<LeaveRoomTask.Params, Unit> {
 
 internal class DefaultLeaveRoomTask @Inject constructor(
         private val roomAPI: RoomAPI,
-        private val eventBus: EventBus
+        private val eventBus: EventBus,
+        private val stateEventDataSource: StateEventDataSource,
+        private val roomSummaryDataSource: RoomSummaryDataSource,
+        private val roomChangeMembershipStateDataSource: RoomChangeMembershipStateDataSource
 ) : LeaveRoomTask {
 
     override suspend fun execute(params: LeaveRoomTask.Params) {
-        return executeRequest(eventBus) {
-            apiCall = roomAPI.leave(params.roomId, mapOf("reason" to params.reason))
+        leaveRoom(params.roomId, params.reason)
+    }
+
+    private suspend fun leaveRoom(roomId: String, reason: String?) {
+        val roomSummary = roomSummaryDataSource.getRoomSummary(roomId)
+        if (roomSummary?.membership?.isActive() == false) {
+            Timber.v("Room $roomId is not joined so can't be left")
+            return
+        }
+        roomChangeMembershipStateDataSource.updateState(roomId, ChangeMembershipState.Leaving)
+        val roomCreateStateEvent = stateEventDataSource.getStateEvent(
+                roomId = roomId,
+                eventType = EventType.STATE_ROOM_CREATE,
+                stateKey = QueryStringValue.NoCondition
+        )
+        // Server is not cleaning predecessor rooms, so we also try to left them
+        val predecessorRoomId = roomCreateStateEvent?.getClearContent()?.toModel<RoomCreateContent>()?.predecessor?.roomId
+        if (predecessorRoomId != null) {
+            leaveRoom(predecessorRoomId, reason)
```

Comment:
If it fails for some reason, it should maybe not be fatal?

Reply:
I start from the predecessors,so I think it's ok